main.go

```diff
@@ -209,97 +227,60 @@ func main() {
+	t := time.NewTicker(opts.Period)
+
+	level.Info(l).Log("msg", "start querying for metrics")
+
+	for {
 		select {
-		case <-ctx.Done():
-			return nil
-		case <-time.After(opts.InitialQueryDelay + opts.Period):
-		}
-		var (
-			rCtx    context.Context
-			rCancel context.CancelFunc
-		)
-		level.Info(l).Log("msg", "start querying for metrics")
-		for {
-			rCtx, rCancel = context.WithCancel(context.Background())
-			select {
-			case <-t.C:
-				if err := read(rCtx, opts.ReadEndpoint, opts.Labels, -1*opts.InitialQueryDelay, opts.Latency, m); err != nil {
-					m.queryResponses.WithLabelValues("error").Inc()
-					level.Error(l).Log("msg", "failed to query", "err", err)
-				} else {
-					m.queryResponses.WithLabelValues("success").Inc()
-				}
-				rCancel()
-			case <-ctx.Done():
-				select {
-				case <-rCtx.Done():
-				case <-time.After(gracefulStopDuration):
-				}
-				rCancel()
-				return reportResults(l, m.queryResponses, opts.SuccessThreshold)
+		case <-t.C:
+			// Do not propagate parent context to prevent cancellation of in-flight request.
+			if err := read(context.Background(), opts.ReadEndpoint, opts.Labels, -1*opts.InitialQueryDelay, opts.Latency, m); err != nil {
+				m.queryResponses.WithLabelValues("error").Inc()
+				level.Error(l).Log("msg", "failed to query", "err", err)
+			} else {
+				m.queryResponses.WithLabelValues("success").Inc()
 			}
+		case <-ctx.Done():
+			t.Stop()
+			return reportResults(l, m.queryResponses, opts.SuccessThreshold)
 		}
-	}, func(_ error) {
-		t.Stop()
-		cancel()
-	})
+	}
 }
 
-func scheduleWriter(g run.Group, opts options, logger log.Logger, m metrics) {
+func runWriter(ctx context.Context, opts options, logger log.Logger, m metrics) error {
 	l := log.With(logger, "component", "writer")
 	t := time.NewTicker(opts.Period)
-	ctx, cancel := context.WithCancel(context.Background())
-
-	g.Add(func() error {
-		level.Info(l).Log("msg", "starting the writer")
-		var (
-			wCtx    context.Context
-			wCancel context.CancelFunc
-		)
-		for {
-			wCtx, wCancel = context.WithCancel(context.Background())
-			select {
-			case <-t.C:
-				if err := write(wCtx, opts.WriteEndpoint, opts.Token, generate(opts.Labels), l); err != nil {
-					m.remoteWriteRequests.WithLabelValues("error").Inc()
-					level.Error(l).Log("msg", "failed to make request", "err", err)
-				} else {
-					m.remoteWriteRequests.WithLabelValues("success").Inc()
-				}
-				wCancel()
-			case <-ctx.Done():
-				level.Info(l).Log("msg", "gracefully stopping writer")
-				select {
-				case <-wCtx.Done():
-				case <-time.After(gracefulStopDuration):
-				}
-				wCancel()
-				return reportResults(l, m.remoteWriteRequests, opts.SuccessThreshold)
+
+	level.Info(l).Log("msg", "starting the writer")
+
+	for {
+		select {
+		case <-t.C:
+			// Do not propagate parent context to prevent cancellation of in-flight request.
+			if err := write(context.Background(), opts.WriteEndpoint, opts.Token, generate(opts.Labels), l); err != nil {
```

Comment:
Now the requests are not cancellable at all, which is not what we want :/ also if we always pass `context.Background()` and not a cencellable ctx why do we even pass the context to this at all?

Reply:
We don't want them to record an error when they are cancelled knowingly. So I believe we can afford to wait for a single in-flight request to complete without cancellation. I tried it with a time-out value, but then I decided to simplify it. I'll have another look at it, and try to recognize these type of scenarios.

For the second part, we can actually get rid of that additional context.

Reply:
hmm I disagree here. the request could hang for as long as the http client timeout is configured, which can be more than a period. Also, I think that intentionally cancelled requests should be counted in the errors as they were requests that did not finish in time.

It should be:
make request
SIGINT
wait for request to finish
if request finishes in time then great
if request does not finish in the grace period then cancel it and count the request as an error